examples/with-next/app/msw-provider.tsx

```diff
@@ -0,0 +1,40 @@
+'use client'
+
+import { Suspense, use } from 'react'
+
+const mockingEnabledPromise =
+  typeof window !== 'undefined'
+    ? import('../mocks/browser').then(async ({ worker }) => {
+        await worker.start({
+          onUnhandledRequest(request, print) {
+            if (request.url.includes('_next')) {
+              return
+            }
+            print.warning()
+          },
+        })
+      })
+    : Promise.resolve()
+
+export function MSWProvider({
+  children,
+}: Readonly<{
+  children: React.ReactNode
+}>) {
```

Comment:
```suggestion
export default ({
  children,
}: Readonly<{
  children: React.ReactNode
}>) => {
```

@kettanaito 

Some success 😄 

The reason the interval is cleared in the Svelte example is because [this listener for `beforeunload`](https://github.com/mswjs/msw/blob/6702d778c541ebae09d5799490f645b57a12c01f/src/browser/setupWorker/start/createStartHandler.ts#L73) is called.

In the Next example, this event listener is not called (presumably `beforeunload` isn't called because of how Next.js works). Therefore, nothing clears the intervals and the workers are not garbage collected.

I'm not 100% sure how it works, since in Svelte `beforeunload` is happening but it doesn't seem like `window.location.reload` is being called. I did notice this meant that changing the handler used in the server rendered component doesn't cause it to be reloaded.

I believe Next.js doesn't perform a full page reload because the MSWProvider has a parent (layout.tsx) so a Fast Refresh is triggered. In a normal React setup, the changed file is at the root (pre-rendering) and has no parents, so a full page refresh is triggered.

We can take advantage of the fact that Fast Refresh use the casing of exports in a file to check if hot reload is possible. See this docs page https://nextjs.org/docs/messages/fast-refresh-reload

Exporting a component as an anonymous component breaks this, and so it always triggers a full reload. Therefore, we can make the change here to get HMR (via full page reload).

I think given this info, it's worth still thinking about `module.hot.dispose()` which does **not** force a full page reload. This mimics the behaviour of the other frameworks.

In my eyes it depends upon if it's a bug or not a bug that `beforeunload` isn't being run in Next.js when there's HMR for the MSWProvider. If that isn't a bug, then there's no reason for setInterval to be cleared and so no reason for the extra worker to disappear. As a relevant note, Svelte mentions [something along these lines](https://github.com/sveltejs/svelte-hmr/blob/master/packages/svelte-hmr/README.md?plain=1#L145) in their HMR package. But at the moment I don't see a massive amount of difference between using `beforeunload` here to do this vs `module.hot.dispose()`

Reply:
I tried both approaches, named and anonymous export, but both fails on: 
[MSW] Warning: intercepted a request without a matching request handler:

In browser logs I clearly see that the worker has started.
```
Starting MSW worker
[MSW] Mocking enabled.
```

On server side the requests are mocked, but on client side all requests are just warning logged.

What I am missing?  Using Next 14.2

My provider looks like this: 
```
'use client';

import { Suspense, use } from 'react';

const mockingEnabledPromise =
  typeof window !== 'undefined' && process.env.NEXT_RUNTIME !== 'nodejs'
    ? import('../mocks/browser').then(async ({ worker }) => {
        console.log('Starting MSW worker');
        await worker.start({
          onUnhandledRequest(request, print) {
            if (request.url.includes('_next')) {
              return;
            }
            const excludedRoutes = [
              'cognito-idp.eu-west-1.amazonaws.com',
              'cognito-identity.eu-west-1.amazonaws.com',
              'google-analytics.com',
            ];
            const isExcluded = excludedRoutes.some((route) => {
              return request?.url?.includes(route);
            });
            if (isExcluded) {
              return;
            }
            print.warning();
          },
        });
      })
    : Promise.resolve();

export default ({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) => {
  return (
    <Suspense fallback={null}>
      <MSWProviderWrapper>{children}</MSWProviderWrapper>
    </Suspense>
  );
};

function MSWProviderWrapper({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  use(mockingEnabledPromise);
  console.log('MSWProviderWrapper children', children);
  return children;
}
```

Reply:
@funes79 not sure this is relevant to this comment I'm afraid. seems like an issue with how you've set up the handlers.

@kettanaito did you see my comment above?

Reply:
I am not sure what issues you are trying to solve here. 

The current state of this PR gives you client-side mocking. If something is not mocked, follow the [Debugging runbook](https://mswjs.io/docs/runbook), the issue is likely elsewhere. 

The only thing missing in the current state is a proper HMR support for client-side mocking.

Reply:
Are you responding to funes? My original comment is about client side HMR

Reply:
@sebws, yes, that is correct. 

Thanks for diving deeper into the HMR issue. 

> Therefore, nothing clears the intervals and the workers are not garbage collected.

So you are saying that interval keeps the `worker` object (not the service worker, mind) persist between HMR? That still sounds a bit odd to me. Do you have proof that clearing that interval indeed solves the issue? 

I don't see `module.hot.dispose()` as the way forward, to be frank. This is a low-level hackery that an average Next.js user shouldn't be exposed to. I don't want to ask developers to do that. The issue is clearly specific to how Next.js handlers client-side HMR, otherwise we had the same issue in other frameworks. This is just a mention that the proper fix for this one is likely on Next.js' side, and that's why we have https://github.com/vercel/next.js/issues/69098.

Reply:
looking again, it looks like more generally it's about `beforeunload` being called and what that does, in particular deregistering the actual service worker. 


```ts
      context.events.addListener(window, "beforeunload", () => {
        if (worker.state !== "redundant") {
          context.workerChannel.send("CLIENT_CLOSED");
        }
        window.clearInterval(context.keepAliveInterval);
      });
```

it's not just the interval which is keeping the `worker` object in memory. we can prove this by logging the creation of the interval, clearing it, and the issue persists.

I may have made the original post at 2am and I don't have a full picture of how msw works so some of this is best guess

Reply:
for context on how I'm looking at this for if you or others are interested, I'm using the "Memory" tab of chrome devtools, and taking heap snapshots at various points. then in the filter, adding `SetupWorkerApi` and looking at the retainers.

I found this pretty confusing but key for me has been clicking through when there is a line of code referenced, as well as trying out what happens when you right click -> ignore this retainer.